steel.1
.\" Manpage for steel.
.\" Any errors or typos, contact niko@byteptr.com.

.TH man 1 "18 Oct 2015" "1.0" "steel man page"
.SH NAME
steel \- command line password manager
.SH SYNOPSIS
steel [OPTIONS]
.SH DESCRIPTION
Steel is a command line password manager for Unix like 
operating systems.
.SH OPTIONS
.IP "-i, --init-new <path>"
Create a new database
.IP "-o, --open <path>"
Open an existing database
.IP "-c, --close"
Close open database
.IP "-a, --add <title> <user> <url> <notes>"
Add new entry to database
.IP "-g, --gen-pass <length> [count]"
Generate secure password
.IP "-d, --delete <id>"
Delete an entry by id
.IP "-r, --replace <id> <what> [content]"
Replace an entry data. <what> can be either
"user", "title", "url", "notes" or "passphrase".
.IP "-R, --shred-db <path>"
Shred database
.IP "-f, --find <search>"
Search database
.IP "-l, --list-all"
Show all entries
.IP "-S, --show-status"
Show database statuses
.IP "-b, --backup <source> <destination>"
Backup database
.IP "-B, --import-backup <source> <destionation>"
Import database backup
.IP "-V, --version"
Show program version
.IP "-p, --show-passphrase <id>"
Show an entry passphrase
.IP "-u, --show-username <id>"
Show an entry username
.IP "-U, --show-url <id>"
Show an entry url
.IP "-n, --show-notes <id>"
Show an entry notes
.IP "-h, --help"
Show short help and exit.
.SH EXAMPLES
Create a new database:
       steel --init-new "/path/to/file.db"
.PP        
Open and decrypt database:
       steel --open "/path/to/existing/file.db"
.PP
Close and encrypt database:
       steel --close
Steel knows what database is currently open and encrypts it.
Close will ask you to type a master passphrase which is used for encryption.
.PP
Add an entry to an open database:
       steel --add "My new entry" "My username" "Url" "Some important notes"
All fields are optional except the title field.
If you don't to insert for example an url, just empty quotes "".
.PP
Replace url in an entry:
       steel --replace 4 "url" "http://www.newurl.com"
.PP
Display on passphrase of an existing entry:
       steel --show-passphrase 4
.PP
Remove database permanently:
       steel --shred-db "/path/to/existing/file.db"
It's not possible to recover shredded database, use with caution.
.PP
Copy passphrase to X-Window system clipboard:
       steel --show-passphrase 4 | xclip -selection clipboard
You will need to install xclip, if it's not already installed on your system.
.PP
If you want to export all entries to a file:
       steel --list-all > file.txt
.SH NOTES
Steel does not have a concept of "change master passphrase". When you close (encrypt)
an open database using --close you can type a master passphrase. This passphrase
is then  required to open (decrypt) the database. You can change the master passphrase
everytime when you close the database, if you want to.
.PP
Note that while using xclip example above might be useful, the passphrase will
be in you clipboard as plain text.
.PP
Steel includes -R option which shreds database files. Note that shredding is not
effective on SSD disks. File will be removed, but not securely.
.SH FILES
.I $HOME/.steel_open
.I $HOME/.steel_dbs
.SH AUTHORS
Written by Niko Rosvall.
.SH COPYRIGHT
Copyright (C) 2015 Niko Rosvall <niko@byteptr.com>
.PP
Released under license GPL-3+. For more information, see
http://www.gnu.org/licenses.


NEWS
1.0: 2015-10-20

    Version 1.0 released.

    This is a very important security fix release. Everyone must upgrade.

    Before upgrading make sure you decrypt (open) your existing database.
    Otherwise you will not be able to open it anymore.

    This releases fixes a security flaw where encryption key was derived by
    SHA256 algorithm only. Version 1.0 now uses bcrypt algorithm together with
    SHA256. This release also uses better seed for generating random
    passwords. (-g, --gen-pass).


0.9: 2015-10-18
	Version 0.9 of Steel released.
	This the initial public stable release of command line password manager 
	Steel. Source is available on Gitlab under the GPL v3 (or later) license.

steel.c
/*
 * Copyright (C) 2015 Niko Rosvall <niko@byteptr.com>
 *
 * This file is part of Steel.
 *
 * Steel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Steel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Steel.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <getopt.h>
#include "cmd_ui.h"

#define VERSION 1.0

static void version_print()
{
	char *str = "This is free software; see the source for copying conditions.\n" \
	"There is no warranty; not even for MERCHANTABILITY or FITNESS\n" \
	"FOR A PARTICULAR PURPOSE.\n";

	printf("Steel v%.1f Copyright (c) Niko Rosvall <niko@byteptr.com>\n",
	       VERSION);
	printf("%s", str);
}

static void usage()
{
#define HELP "\
SYNOPSIS\n\
\n\
steel [options]\n\
\n\
OPTIONS\n\
\n\
-i, --init-new          <path>                        Create a new database\n\
-o, --open              <path>                        Decrypt existing database\n\
-c, --close                                           Encrypt open database\n\
-a, --add               <title> <user> <url> <notes>  Add new entry to database\n\
-s, --show              <id>                          Show entry by id\n\
-g, --gen-pass          <length> [count]              Generate secure password\n\
-d, --delete            <id>                          Delete an entry by id\n\
-r, --replace           <id> <what> [content]         Replace an entry data\n\
						      <what> can be either \"user\",\n\
						      \"title\", \"url\", \"notes\" or\n\
					              \"passphrase\".\n\
-R, --shred-db          <path>                        Shred database\n\
-f, --find              <search>                      Search database\n\
-l, --list-all                                        Show all entries\n\
-S, --show-status                                     Show database statuses\n\
-b, --backup            <source> <destination>        Backup database\n\
-B, --import-backup     <source> <destination>        Import database backup\n\
-V, --version                                         Show program version\n\
-p, --show-passphrase   <id>                          Show an entry passphrase\n\
-u, --show-username     <id>                          Show an entry username\n\
-U, --show-url          <id>                          Show an entry url\n\
-n, --show-notes        <id>                          Show an entry notes\n\
-h, --help                                            Show short help and exit.\n\
\n\
For more information and examples see man steel(1).\n\
\n\
AUTHORS\n\
Copyright (C) 2015 Niko Rosvall <niko@byteptr.com>\n\
\n\
Released under license GPL-3+. For more information, see\n\
http://www.gnu.org/licenses\n\
"
	printf(HELP);
}

//Program entry point.
int main(int argc, char *argv[])
{
	int option;

	if(argc == 1) {
		usage();
		return 0;
	}

	while(true) {

		static struct option long_options[] =
		{
			{"init-new",               required_argument, 0, 'i'},
			{"backup",                 required_argument, 0, 'b'},
			{"import-backup",          required_argument, 0, 'B'},
			{"open",                   required_argument, 0, 'o'},
			{"close",                  no_argument,       0, 'c'},
			{"show",                   required_argument, 0, 's'},
			{"gen-pass",               required_argument, 0, 'g'},
			{"add",                    required_argument, 0, 'a'},
			{"delete",                 required_argument, 0, 'd'},
			{"replace",                required_argument, 0, 'r'},
			{"shred-db",               required_argument, 0, 'R'},
			{"find",                   required_argument, 0, 'f'},
			{"list-all",               no_argument,       0, 'l'},
			{"show-status",            no_argument,       0, 'S'},
			{"version",                no_argument,       0, 'V'},
			{"help",                   no_argument,       0, 'h'},
			{"show-passphrase",        required_argument, 0, 'p'},
			{"show-username",          required_argument, 0, 'u'},
			{"show-url",               required_argument, 0, 'U'},
			{"show-notes",             required_argument, 0, 'n'},
			{0, 0, 0, 0}

		};

		int option_index = 0;

		option = getopt_long(argc, argv, "i:b:B:o:cs:g:a:d:r:f:lR:SVp:u:U:n:h",
				     long_options, &option_index);

		if(option == -1)
			break;

		switch(option) {
		case 'i':
			init_database(optarg);
			break;
		case 'b':
			if(!argv[optind]) {
				fprintf(stderr, "Missing option <destination>.\n");
				return 0;
			}
			backup_database(optarg, argv[optind]);
			break;
		case 'B':
			if(!argv[optind]) {
				fprintf(stderr, "Missing option <destination>.\n");
				return 0;
			}
			backup_import_database(optarg, argv[optind]);
			break;
		case 'o':
			open_database(optarg);
			break;
		case 'c':
			close_database();
			break;
		case 's':
			show_one_entry(atoi(optarg));
			break;
		case 'g': {
			int count = 1;

			if(argv[optind]) {
				count = atoi(argv[optind]);

				if(count < 1)
					count = 1;
			}

			generate_password(atoi(optarg), count);
			break;
		}
		case 'a': {
			//Has user?
			if(!argv[optind]) {
				fprintf(stderr, "Missing option user.\n");
				return 0;
			}
			//Has url?
			if(!argv[optind + 1]) {
				fprintf(stderr, "Missing option url.\n");
				return 0;
			}
			//Has note?
			if(!argv[optind + 2]) {
				fprintf(stderr, "Missing option note\n");
				return 0;
			}

			char *title = optarg;
			char *user = argv[optind];
			char *url = argv[optind + 1];
			char *note = argv[optind + 2];

			add_new_entry(title, user, url, note);

			break;
		}
		case 'd':
			delete_entry(atoi(optarg));
			break;
		case 'r': {
			if(!argv[optind]) {
				fprintf(stderr, "Missing option, see -h for help\n");
				return 0;
			}

			//Replacing passphrase does not need third argument
			//It will be asked separately by replace_part()
			if(strcmp(argv[optind], "passphrase") != 0) {
				if(!argv[optind + 1]) {
					fprintf(stderr, "Missing option, see -h for help\n");
					return 0;
				}
			}

			int id = atoi(optarg);
			char *what = argv[optind];
			char *content = argv[optind + 1];
			replace_part(id, what, content);
			break;
		}
		case 'f':
			find_entries(optarg);
			break;
		case 'p':
			show_passphrase_only(atoi(optarg));
			break;
		case 'u':
			show_username_only(atoi(optarg));
			break;
		case 'U':
			show_url_only(atoi(optarg));
			break;
		case 'n':
			show_notes_only(atoi(optarg));
			break;
		case 'l':
			show_all_entries();
			break;
		case 'R':
			remove_database(optarg);
			break;
		case 'S':
			show_database_statuses();
			break;
		case 'V':
			version_print();
			break;
		case 'h':
			usage();
			break;
		}

	}

	return 0;
}
